### PR TITLE
[highlightjs] Add new addon

### PR DESCRIPTION
**Pull request by MrPetovan**
Related to https://github.com/friendica/friendica/pull/5765

Since we aren't doing syntax highlighting in the core, I'd figured people would like a replacement, and this is the one.

Depends on https://github.com/friendica/friendica/pull/5773

### CODE REVIEW

**Comment by annando** [comment]
We are so late in the release that I really have problems in merging some huge PR like this and the corresponding one in the core. We surely would need several days of testing - but I would prefer having the release this week.

**Comment by MrPetovan** [comment]
This can actually wait, the other can't. No post with highlighted code was coming through at all.

**Comment by rabuzarus** [comment]
Does it work with new content added by ajax (next page / new posts) ?

**Comment by MrPetovan** [comment]
That's an excellent question, I haven't been able to see this in the wild yet, but I don't believe it would.

**Comment by MrPetovan** [comment]
Issue confirmed, no syntax highlighting on ajax (re)load.

**Comment by rabuzarus** [comment]
For new posts:
I would say we would need a new js hook in `main.js` before we update the scroll position. Maybe at the end of `updateConvItems()`. When the hook is called the addon would need to execute something like
```
$('pre > code').each(function() {
     hljs.highlightBlock(this);
})
```
(https://github.com/highlightjs/highlight.js/issues/966#issuecomment-152122946)
or
```
var blocks = document.querySelectorAll('pre code:not(.hljs)');
Array.prototype.forEach.call(blocks, highlightBlock);)
```
(https://github.com/highlightjs/highlight.js/issues/909#issuecomment-131686186)

Same for `loadScrollContent();`

or we use another javascript code highlighter

**Comment by rabuzarus** [comment]
> or we use another javascript code highlighter

maybe PRISM (http://prismjs.com). We would execute `Prism.highlightAll();` at the end of `updateConvItems()` (https://schier.co/blog/2013/01/07/how-to-re-run-prismjs-on-ajax-content.html) or if we want to have an eye on performance we could use `Prism.highlightAllUnder(element, async, callback)` for the new content (https://prismjs.com/extending.html).

**Comment by MrPetovan** [comment]
Thank you for looking into it, my plan was to add an custom event on the `<body>` tag triggered by the Ajax load and then add an event listener in the `<script>` tag added by the addon. This is the most unobtrusive way of handling this, and it may be used for other addons as well without having to tamper with the main JS routine.

**Comment by MrPetovan** [comment]
Phew!

I managed to make syntax highlighting work with Ajax, but not without completely reworking the Javascript hooks and rewriting the MathJax addon in the process 🤔

File: highlightjs/highlightjs.php [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	$a->page['htmlhead'] .= <<< HTML
+
+<link rel="stylesheet" href="{$basedir}/styles/{$style}.css">
+
+HTML;
+
+	$b .= <<< HTML
+
+<script type="text/javascript" src="{$basedir}/highlight.pack.js"></script>
+<script type="text/javascript">
```
Comment (rabuzarus) [inline highlightjs/highlightjs.php]: I would prefer to put the following javascript code into an own js file. We should avoid to mix php and js if it is possible

Comment (MrPetovan) [reply highlightjs/highlightjs.php]: They are cached, but the first load is more expensive when it doesn't necessarily has to be.

I'll see what I can do.

**Comment by MrPetovan** [comment]
I added ways to register stylesheets and javascript files so that no HTML is written by hand in the hooks.

I had to create two new hooks and remove a completely useless part of `App->page`, but it's ready now.

**Pull request merged**